Model/AdyenThreeDS2Process.php

```diff
@@ -118,24 +133,35 @@ public function initiate($payload)
             return $this->adyenHelper->buildThreeDS2ProcessResponseJson($result['resultCode'], $result['authentication']['threeds2.challengeToken']);
         }
 
-        // Payment can get back to the original flow
-
         // Save the payments response because we are going to need it during the place order flow
         $payment->setAdditionalInformation("paymentsResponse", $result);
 
-        // Setting the placeOrder to true enables the process to skip the payments call because the paymentsResponse
-        // is already in place - only set placeOrder to true when you have the paymentsResponse
-        $payment->setAdditionalInformation('placeOrder', true);
-
         // To actually save the additional info changes into the quote
-        $quote->save();
+        $order->save();
 
-        $this->adyenLogger->addAdyenDebug("CC payment finished for order: " . $quote->getReservedOrderId());
-        if (!empty($result['resultCode'])) {
-            $this->adyenLogger->addAdyenDebug('Result code: ' . $result['resultCode']);
+
+        $response = [];
+
+        if($result['resultCode'] != 'Authorised') {
+            $this->checkoutSession->restoreQuote();
+            throw new \Magento\Framework\Exception\LocalizedException(__('The payment is REFUSED.'));
         }
 
-        // 3DS2 flow is done, original place order flow can continue from frontend
-        return $this->adyenHelper->buildThreeDS2ProcessResponseJson();
+        $response['result'] = $result['resultCode'];
+        return json_encode($response);
+    }
+
+    /**
+     * Get order object
+     *
+     * @return \Magento\Sales\Model\Order
+     */
+    protected function getOrder()
+    {
+        if (!$this->order) {
+            $incrementId = $this->checkoutSession->getLastRealOrderId();
```

Comment:
if we get the last real order id but it has ben already restored then what is gonna happen to the cancel notificitation which is sent(?) after the first attemt was refused for example?

Reply:
I would assume we do a new reserver order id but let's test this.

Reply:
added to the ticket